phat/crates/rollup/src/clients/ink.rs

```diff
@@ -0,0 +1,424 @@
+use alloc::vec::Vec;
+
+use ink::primitives::Hash;
+use kv_session::{
+    rollup,
+    traits::{
+        BumpVersion, Key, KvSession, KvSnapshot, QueueIndex, QueueIndexCodec, QueueSession, Value,
+    },
+    RwTracker, Session,
+};
+use pink_extension::chain_extension::signing;
+use pink_extension::ResultExt;
+
+#[cfg(feature = "logging")]
+use pink_extension::debug;
+
+use primitive_types::H256;
+use scale::{Decode, Encode};
+use subrpc::contracts::*;
+
+pub use crate::{Action, Error, Result};
+
+const DEFAULT_QUEUE_PREFIX: &[u8] = b"q/";
+
+pub type ContractId = [u8; 32];
+
+pub struct InkSnapshot<'a> {
+    rpc: &'a str,
+    pallet_id: u8,
+    call_id: u8,
+    contract_id: &'a ContractId,
+    at: H256,
+}
+
+impl<'a> InkSnapshot<'a> {
+    pub fn new(
+        rpc: &'a str,
+        pallet_id: u8,
+        call_id: u8,
+        contract_id: &'a ContractId,
+    ) -> Result<Self> {
+        let hash = subrpc::get_block_hash(rpc, None).or(Err(Error::FailedToGetBlockHash))?;
+        Ok(InkSnapshot {
+            rpc,
+            pallet_id,
+            call_id,
+            contract_id,
+            at: hash,
+        })
+    }
+}
+
+impl<'a> KvSnapshot for InkSnapshot<'a> {
+    fn get(&self, key: &[u8]) -> kv_session::Result<Option<Value>> {
```

Comment:
Maybe in the next PR, consider adding a "get_batch" implementation for better speed